lib/internal/streams/operators.js

```diff
@@ -0,0 +1,148 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const Readable = require('internal/streams/readable');
+const eos = require('internal/streams/end-of-stream');
+const { isNodeStream, isReadableNodeStream } = require('internal/streams/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  },
+  AbortError,
+} = require('internal/errors');
+
+module.exports.map = function map(stream, fn, options) {
+  let concurrency = 1;
+  if (options && options.concurrency != null) {
+    concurrency = options.concurrency;
+  } else if (options != null && typeof options !== 'object') {
+    concurrency = options;
+  }
+
+  if (!Number.isInteger(concurrency)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'concurrency', ['Integer'], concurrency);
+  }
+
+  if (!isNodeStream(stream) && isIterable(stream)) {
+    // TODO (perf): Add support for Iterable and AsyncIterable in pump.
+    stream = Readable.from(stream);
+  }
+
+  if (!isReadableNodeStream(stream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'stream', ['Readable', 'Iterable', 'AsyncIterable'], stream);
+  }
+
+  if (typeof fn !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'fn', ['Function', 'AsyncFunction'], stream);
+  }
+
+  const ac = new AbortController();
+  const signal = ac.signal;
+  const queue = [];
+
+  let reading = false;
+  let errored = false;
+  let ended = false;
+
+  const ret = new Readable({
+    objectMode: stream.readableObjectMode ?? stream.objectMode ?? true,
+    highWaterMark: 0,
+    signal: options && options.signal,
```

Comment:
Should this default on the signal above if the options aren't passed? (Also, can use options?.signal)